File: PA3/pa-3.py
import os
import random
import timeit
import math
from decimal import Decimal  # used for scientific notation

# FIB_NUMS = [19, 20, 30, 35, 40, 45, 50, 55]
FIB_NUMS = [19, 20, 30]

FILENAME = "Fibonacci_Time.csv"
HEADERS = ["T1: Recursive Algorithm (ms)", "T2: DP Algorithm (ms)", "(2^n)/n", "T1/T2"]


def main():
    output_to_csv(FILENAME, HEADERS, run_tests())
    print("Done.\nFile saved to " + os.getcwd() + "\\" + FILENAME)
    console_interface()


########################################################################################################################
#  Classes
########################################################################################################################
class TestResults(object):
    def __init__(self, time_rec, time_dp, n):
        self.time_rec = time_rec
        self.time_dp = time_dp
        self.val1 = math.pow(2, n) / 2
        self.val2 = '%0.E' % Decimal(self.time_rec / self.time_rec)

    def to_string(self):
        return str(self.time_rec), str(self.time_dp), str(self.val1), str(self.val2)


########################################################################################################################
#  Fibonacci
########################################################################################################################
# Recursive
def recur_fib(num):
    if num <= 1:
        return num
    return recur_fib(num - 1) + recur_fib(num - 2)


# temp place holder
def dp(num):
    if num <= 1:
        return num
    return recur_fib(num - 1) + recur_fib(num - 2)


########################################################################################################################
#  Time Execution Test
########################################################################################################################
# Measures execution time of each merge sort and saves results to an array of MergeSortResult objects
def calc_time(alg, num):
    start_time = timeit.default_timer()  # start time measure
    fib_num = alg(num)
    stop_time = timeit.default_timer()  # end time measure
    return (stop_time - start_time) * 1000


def run_tests():
    results = []
    for i in FIB_NUMS:
        results.append(
            TestResults(
                calc_time(recur_fib, i),
                calc_time(dp, i),
                i
            )
        )
    return results


########################################################################################################################
#  Output Results
########################################################################################################################
# csv
def output_to_csv(filename, headers, data):
    with open(filename, 'w') as file:
        for header in headers:
            file.write(header)
            file.write(",")
        file.write("\n")
        for d in data:
            file.write(",".join(d.to_string()))
            file.write("\n")


# interface for user to request sorted and unsorted arrays of the test runs
def console_interface():
    while True:
        print("\nSelect Recursive(R) or Dynamic(D) Fibonacci Algorithm or q to quit\n")
        user_in = input("->  ")
        try:
            if user_in == 'q':
                print("Exiting ... \n")
                break
            elif user_in == 'R':
                print("\nEnter a Number\n")
                user_in = input("->  ")
                print(recur_fib(int(user_in)))
            elif user_in == 'D':
                print("\nEnter a Number\n")
                user_in = input("->  ")
                print(dp(int(user_in)))
        except:
            print("Invalid input: ", user_in, "\n")


########################################################################################################################
#  Helpers
########################################################################################################################
# Help function which creates the random array of test inputs
def create_random_array(array_length, rand_min, rand_max):
    rand = []
    for num in range(array_length):
        rand.append(random.randint(rand_min, rand_max))
    return rand


########################################################################################################################
#  Run Script
########################################################################################################################
if __name__ == "__main__":
    main()

File: PA3/PA3.java
// Robert Allen, William Blackwell, Bradley Sutton
// Implements recursive and dynamic programming versions of Fibonacci algorithm

import helperclasses.Algorithm;
import helperclasses.AlgorithmResult;
import helperclasses.TestResult;

import java.io.IOException;
import java.io.PrintWriter;
import java.util.*;

public class PA3 {

    // Values assigned for testing
    private static final int[] FIB_TEST_VALUES = {10, 20, 30, 35, 40, 45, 50, 55};

    // Used by dynamic programming algorithm.  Allows for 'memorization' of Fib values
    private static final Map<Integer, Long> DP_MAP = new HashMap<>();

    // CSV info
    private static final String FILENAME = "Fibonacci_Time.csv";
    private static final String[] HEADERS = {
            "n", "F(n)", "T1: Recursive Algorithm (ns)/(s)/or(m)", "T2: DP Algorithm (ns)", "(2^n)/n", "T1/T2"};

    // Run Program
    public static void main(String[] args) {
        //initialize DP_MAP
        initDPArray();

        // Start interactive console with user
        fibConsole();

        // Ask user if they would like to run the tests in the given PA.  Warn the user that some values take
        // a very long time to process.
        System.out.println("\nPerform new calculations for the assigned PA values and save to file?\nAssigned Values: " +
                Arrays.toString(FIB_TEST_VALUES) + "\nCAUTION! Takes LONG Time [y/n]  ");

        // Process user choice
        String userChoice = new Scanner(System.in).next();
        if (userChoice.equalsIgnoreCase("Y")) { // yes save
            // Run Tests
            ArrayList<TestResult> results = runTests();
            try {
                writeToCSV(results);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        // Exit
        else { // no
            System.out.println("Exiting");
        }
    }

    // Measure execution time in (ns)
    private static AlgorithmResult calcTime(Algorithm fibAlg, int n) {
        long start = System.nanoTime();
        long result = fibAlg.algorithm(n);
        long stop = System.nanoTime();
        return new AlgorithmResult(result, stop - start);
    }

    // Run tests using the recursive alg and the dp alg
    private static ArrayList<TestResult> runTests() {
        ArrayList<TestResult> results = new ArrayList<>();
        for (int i : FIB_TEST_VALUES) {
            AlgorithmResult recursiveTest = calcTime(PA3::fibRecursive, i);
            AlgorithmResult dynamicTest = calcTime(PA3::fibDynamic, i);
            results.add(new TestResult(dynamicTest.getFibNum(), recursiveTest.getTime(), dynamicTest.getTime(), i));
        }
        return results;
    }

    // Calculates Fibonacci Number using Recursion
    private static long fibRecursive(int n) {
        return n <= 1 ? 1 : fibRecursive(n - 1) + fibRecursive(n - 2);
    }

    // Calculates Fibonacci Number using Dynamic Programming
    // Results of calculations saved to the DP_MAP dictionary
    private static long fibDynamic(int n) {
        // If the key already exists in DP_MAP, the algorithm has already committed it to memory, ro no need to
        // recalculate it.  Otherwise, find F(n)
        if (!DP_MAP.containsKey(n)) for (int i = 2; i <= n; i++) {
            long val = DP_MAP.get(i - 1) + DP_MAP.get(i - 2);
            DP_MAP.put(i, val);
        }
        return DP_MAP.get(n);
    }

    // Output results to csv file
    private static void writeToCSV(ArrayList<TestResult> results) throws IOException {
        PrintWriter writer = new PrintWriter(FILENAME, "UTF-8");
        writer.println(String.join(",", HEADERS));
        for (TestResult result : results) {
            writer.println(result.toString());
        }
        writer.close();
    }

    // Asks the user to select an algorithm to use to fin Fibonacci numbers.  Also displays execution time.
    private static void fibConsole() {
        Scanner scan = new Scanner(System.in);
        while (true) {
            System.out.println("Please enter (R) if you'd like to use the Recursive Fib Algorithm \n" +
                    "or enter (D) if you'd like to use the Dynamic Fib Algorithm \n" +
                    "or enter (Q) to quit.");

            // Process user input
            String algType = scan.next();

            // Quit interactive console
            if (algType.equalsIgnoreCase("Q")) {
                System.out.println("Thank you for using my program!");
                return;
            }

            // Find F(n) used to selected algorithm
            System.out.println("Please enter a positive integer value you'd like the fibonacci number to: ");
            int num = scan.nextInt();

            // Recursion
            if (algType.equalsIgnoreCase("R")) {
                displayResult(num, calcTime(PA3::fibRecursive, num));
            }
            // Dynamic
            else if (algType.equalsIgnoreCase("D")) {
                displayResult(num, calcTime(PA3::fibDynamic, num));
            }
        }
    }

    // Helper function for interactive console
    // Calculates answer from user and shows her/him the result
    private static void displayResult(int num, AlgorithmResult answer) {
        System.out.println("The value for " + num + " is: " + answer.getFibNum());
        double seconds = (double) answer.getTime() / 1_000_000_000.0;
        System.out.println("The time taken to complete the recursive algorithm for valued "
                + num + " was: " + seconds + " seconds.\n\n");
    }

    // Initialize DP_MAP
    private static void initDPArray() {
        DP_MAP.put(0, (long) 1);
        DP_MAP.put(1, (long) 1);
    }
}
